ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -95,14 +115,43 @@ MessageFormatFlags getOperationFlag() {
   @Override
   void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
     //First, check if any of the existing requests have timed out.
-    cleanupExpiredInFlightRequests();
-    checkAndMaybeComplete();
-    if (!isOperationComplete()) {
-      fetchRequests(requestRegistrationCallback);
+    if (!maybeProcessDecryptCallbackAndComplete()) {
```

Comment:
I think it might be cleaner to do this inside `checkAndMaybeComplete()`. You currently call `checkAndMaybeComplete()` from too many places.